torchgeo/trainers/simclr.py

```diff
@@ -0,0 +1,242 @@
+
+    * v1: https://arxiv.org/abs/2002.05709
+    * v2: https://arxiv.org/abs/2006.10029
+
+    .. versionadded:: 0.5
+    """
+
+    def __init__(
+        self,
+        model: str = "resnet50",
+        weights: Optional[Union[WeightsEnum, str, bool]] = None,
+        in_channels: int = 3,
+        version: int = 2,
+        layers: int = 3,
+        hidden_dim: Optional[int] = None,
+        output_dim: Optional[int] = None,
+        lr: float = 4.8,
+        weight_decay: float = 1e-4,
+        max_epochs: int = 100,
+        temperature: float = 0.07,
+        memory_bank_size: int = 64000,
+        gather_distributed: bool = False,
+        augmentations: nn.Module = AUG,
+    ) -> None:
+        """Initialize a new SimCLRTask instance.
+
+        Args:
+            model: Name of the timm model to use.
+            weights: Either a weight enum, the string representation of a weight enum,
+                True for ImageNet weights, False or None for random weights,
+                or the path to a saved model state dict.
+            in_channels: Number of input channels to model.
+            version: Version of SimCLR, 1--2.
+            layers: Number of layers in projection head (2 for v1 or 3+ for v2).
+            hidden_dim: Number of hidden dimensions in projection head
+                (defaults to output dimension of model).
+            output_dim: Number of output dimensions in projection head
+                (defaults to output dimension of model).
+            lr: Learning rate (0.3 x batch_size / 256 is recommended).
+            weight_decay: Weight decay coefficient (1e-6 for v1 or 1e-4 for v2).
+            max_epochs: Maximum number of epochs to train for.
+            temperature: Temperature used in NT-Xent loss.
+            memory_bank_size: Size of memory bank (0 for v1 or 64K for v2).
+            gather_distributed: Gather negatives from all GPUs during distributed
+                training (ignored if memory_bank_size > 0).
+            augmentations: Data augmentation.
+        """
+        super().__init__()
+
+        self.save_hyperparameters(ignore=["augmentations"])
+        self.augmentations = augmentations
+
+        # Create backbone
+        self.backbone = timm.create_model(
+            model, in_chans=in_channels, pretrained=weights is True
+        )
+
+        # Load weights
+        if weights and weights is not True:
+            if isinstance(weights, WeightsEnum):
+                state_dict = weights.get_state_dict(progress=True)
+            elif os.path.exists(weights):
+                _, state_dict = utils.extract_backbone(weights)
+            else:
+                state_dict = get_weight(weights).get_state_dict(progress=True)
+            self.backbone = utils.load_state_dict(self.backbone, state_dict)
+
+        # Create projection head
+        input_dim = list(self.backbone.named_children())[-1][1].out_features
```

Comment:
Most timm models have a `num_features` attr. Should we prefer this over `list(self.backbone.named_children())[-1][1].out_features`?

Reply:
I replaced this because our fask classification model doesn't have this attr, but we could add it. Why didn't we go with `num_features` in the weight loading PR? If this long complicated thing is more reliable we should go with that.

Reply:
I don't think it's more reliable because it probably makes assumptions about the layer you indexed which may not be the case for other timm models